main/src/main/java/com/kenshoo/facebookgateway/entities/CampaignResultRO.java

```
@@ -19,18 +19,26 @@
     private static final ToStringStyle SHORT_PREFIX_STYLE = ToStringStyle.SHORT_PREFIX_STYLE;
 
     private final Collection<CampaignRO> campaignROs;
+    private final String errorMessage;
 
     @JsonCreator
-    public CampaignResultRO(@JsonProperty(value = "campaignROs")
-                            final Collection<CampaignRO> campaignROs) {
+    public CampaignResultRO(@JsonProperty(value = "campaignROs") final Collection<CampaignRO> campaignROs,
```

@lior-gershony refresh my bad memory, is this the object returning to the client (KS/Social?) - I think that for best practice we'd possibly want:
* Internal (Kenshoo) error code? (possibly that should be handled by client and not the Gateway)
* Facebook Error code(s) - If I recall they have a multi (2/3) level error code system, of which some of the codes are embedded in the error message
* any other info we have to add to that

So in any case the error message should probably be expected to be a struct of some kind and not a string.
This can help clients manage/handle errors not yet handled in the Gateway.
WDYT?

---

@mosheeshel first of all - thanks for the review :)
yes, this is the object returning to the client.

1. regarding Facebook error codes, I thought that we want to handle them in the gateway in order to setup the retyer accordingly. (some errors are absolute failures while others are temporary failures and can be retried). In my mind - the gateway should return a failure to the client when the retryer gave up (absolute failure or upper bound of retries), but in any cases I thought that the error message itself is enough in order for the client to understand why he didn't get the required results. and in both cases - Facebook error codes logic is handled in the gateway.
am I missing something?
2. regarding Internal (Kenshoo) error code - not sure what that means.

---

@mosheeshel to not hold up testing - I am merging this and we can take the conversation offline ok?

---

I totally agree, we should handle whatever we can at the Gateway level, no need to return anything to client if we succeed. However, when it doesn't succeed, there can be a few reasons:
1. It can be a fatal error - requests for ads in a campaign that doesn't exist
2. something that usually works after retry - but didn't this time (I think we do need to let client know about this)
3. When we just don't know how to handle - a new type of error, or new object issues (like we get sometimes with FB when they introduce something new without us catching up) - some clients handle by ignoring & notifying client / failing completely...